### PR TITLE
[Merged by Bors] - feat(Mathlib/Tactic/Clear_): add clear_ tactic

### PR DESCRIPTION
**Pull request by JOSHCLUNE**

### CODE REVIEW

File: test/Clear_.lean [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import Mathlib.Tactic.Clear_
+import Mathlib.Tactic.Replace
+
+-- Most basic test
+example (_delete_this : Nat) : Nat := by
+  clear_
+  fail_if_success assumption
+  exact 0
+
```
Comment (digama0) [inline test/Clear_.lean:9]:
```suggestion

-- Confirms that clear_ does not delete class instances
example [_dont_delete_this : Inhabited Nat] : Inhabited Nat := by
  clear_
  assumption

```

File: Mathlib/Tactic/Clear_.lean [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+              return ldecl :: ldecls
+            else
+              return ldecls
+          let dependencies ← lctx.foldrM add_iff_dependent []
+          let goal_decl ← getMVarDecl (← getMainGoal)
+          if dependencies.isEmpty && not (← exprDependsOn goal_decl.type fvarId) then
+            lctx := lctx.erase fvarId
+            let mvarId ← clear (← getMainGoal) fvarId
+            replaceMainGoal [mvarId]
+      | none => continue
```
Comment (digama0) [inline Mathlib/Tactic/Clear_.lean:61]:
```suggestion
/-- Clear all hypotheses starting with `_`, like `_match` and `_let_match`. -/
elab (name := clear_) "clear_" : tactic =>
  liftMetaTactic1 fun goal => do
    let mut toClear := #[]
    for decl in ← getLCtx do
      if let Name.str _ str := decl.userName then
        if !str.isEmpty && str.front == '_' then
          if let none ← isClass? decl.type then
            toClear := toClear.push decl.fvarId
    tryClearMany goal toClear
```

File: test/Clear_.lean [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  assumption
+
+-- Confirms that clear_ does not break the goal
+example (_dont_delete_this : Type) : _dont_delete_this = _dont_delete_this := by
+  clear_
+  rfl
+
+-- Confirms that clear_ clears all that it can even if some underscored hypotheses cannot be cleared
+example (_dont_delete_this : Type) (_delete_this : _dont_delete_this = _dont_delete_this)
+  : _dont_delete_this = _dont_delete_this := by
```
Comment (semorrison) [inline test/Clear_.lean:37]: mathlib style, which I assume will stay the same for mathlib4, is that we preferring trailing `:` over leading `:`.

**Pull request closed without merging**